core/domain/feedback_services.py

```diff
@@ -235,21 +296,54 @@ def update_messages_read_by_the_user(user_id, thread_id, message_ids):
     feedback_thread_user_model.put()
 
 
-def add_message_id_to_read_by_list(thread_id, user_id, message_id):
-    """Adds the message id to the list of message ids read by the user.
+def add_message_ids_to_read_by_list(thread_ids, user_id, message_ids):
+    """Adds the message ids to the list of message ids read by the user.
 
     Args:
-        thread_id: str. The id of the thread.
+        thread_ids: list(str). The ids of the threads.
         user_id: str. The id of the user reading the messages.
-        message_id: int. The id of the message.
-    """
-    feedback_thread_user_model = (
-        feedback_models.GeneralFeedbackThreadUserModel.get(
-            user_id, thread_id) or
-        feedback_models.GeneralFeedbackThreadUserModel.create(
-            user_id, thread_id))
-    feedback_thread_user_model.message_ids_read_by_user.append(message_id)
-    feedback_thread_user_model.put()
+        message_ids: list(str). The ids of the messages.
+    """
+    # First get all of the GeneralFeedbackThreadUserModels that already exist.
+    current_feedback_thread_user_models = (
+        feedback_models.GeneralFeedbackThreadUserModel.get_multi(
+            user_id, thread_ids))
+
+    # Keep track of which thread_ids do not have models yet.
+    missing_thread_ids = []
+    # Keep track of the current_feedback_thread_user_models that aren't None.
+    feedback_thread_user_models = []
+    for index, feedback_model in enumerate(
+            current_feedback_thread_user_models):
+        if feedback_model is None:
+            missing_thread_ids.append(thread_ids[index])
+        else:
+            feedback_thread_user_models.append(feedback_model)
+
+    # Create the new GeneralFeedbackThreadUserModels for each of the thread_ids
+    # that do not have a model yet.
+    new_feedback_thread_user_models = []
+    if len(missing_thread_ids) > 0:
+        new_feedback_thread_user_models = (
+            feedback_models.GeneralFeedbackThreadUserModel.create_multi(
+                user_id, missing_thread_ids)
+        )
+
+    # The GeneralFeedbackThreadUserModels that need to be updated are the new
+    # models that we created as well as the models that already existed.
+    feedback_thread_user_models.extend(new_feedback_thread_user_models)
+
+    # For each of the models, append the message_id to the
```

Comment:
Haven't you changed the order of the elements as a result of the extend() call? So this would lead to an incorrect result?

(If that's the case, please add a test that detects this before fixing the bug.)

Reply:
Instead of creating a list of feedback_thread_user_models in the original order (using insert) I can treat the old models and new models as separate lists and update their message_ids_read_by_user property separately...just another thought :)

Reply:
Yes, that seems reasonable too.

Reply:
In fact, I think that would be cleaner. The current implementation is rather complex, and not that easy to follow...